### DISCUSSION TITLE
Only ~25k RPS when running benchmark via redis-benchmark

### DISCUSSION THREAD

**Discussion by kamyuentse** [microsoft/garnet#68] [Q&A] [answered]
I follow the "Get Started" to build&run GarnetServer,  then execute a benchmark via redis-benchmark. The result that ~25k RPS on my machine. Did I miss something?

command: `redis-benchmark -t set,get -n 10000000 -p 3278 -c 100 -r 10000000`

OS: 20.04.1-Ubuntu
dotnet version: 8.0.202
CPU: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz

<img width="1252" alt="image" src="https://github.com/microsoft/garnet/assets/15538025/43e9f579-05a8-4a85-ae01-b51df841ac80">

**Comment by badrishc** [comment, marked as answer]
1. First, `redis-benchmark` is known to be extremely heavy on the client side, as it was developed for a different tier of performance.
2. It looks like you have not enabled pipelining which means each request is being fully processed and returned before sending the next in a session. In this mode you are measuring nothing but the performance of `localhost` network roundtrip time and TCP overhead. Try `-P 1024` for example.
3. Either use `Resp.benchmark` which is carefully designed to ensure client does not become the bottleneck, or if you want a third-party tool then `memtier_benchmark` ([link](https://github.com/RedisLabs/memtier_benchmark)) is better than `redis-benchmark`.

**Comment by kinthaiofficial** [comment]
~25k RPS via redis-benchmark against Garnet is lower than expected given Garnet's claims of multi-million RPS in internal Microsoft benchmarks. A few things to check:

**1. Pipeline depth** — redis-benchmark defaults to no pipelining, which is latency-bound rather than throughput-bound:
```bash
# Without pipelining: tests latency, not throughput
redis-benchmark -h localhost -p 6379 -c 50 -n 100000 GET mykey

# With pipelining: tests actual throughput
redis-benchmark -h localhost -p 6379 -c 50 -n 100000 -P 32 GET mykey
```

**2. Client count** — default `-c 50` may not saturate Garnet's thread pool:
```bash
redis-benchmark -c 500 -P 32 -n 1000000 GET mykey
```

**3. Network path** — running benchmark on the same host (loopback) vs. cross-network changes results significantly:
```bash
# Same host: should be much higher
# Cross-network: depends on NIC, MTU, etc.
```

**4. Garnet config** — make sure thread count is set to match CPU cores:
```json
// garnet.conf
{
  "ThreadPoolMinThreads": 16,
  "ThreadPoolMaxThreads": 64
}
```

What's your test setup (same machine, cross-network, VM vs. bare metal)? And are you using pipelining? That'll help narrow down whether this is a config issue or a genuine performance ceiling.
